enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,348 @@
+
+* Systemd services of the then optional networking components (like `microshift-ovs-init`)
+  should still start before the `microshift` systemd service, but `microshift` cannot
+  depend on `microshift-ovs-init` anymore. This can be acomplished by using the
+  Before=microshift clause in `microshift-ovs-init.service` and enabling this service
+  during the `microshift-networking-ovnk` sub-package install.
+
+* The rebase script should handle extracting and updating the component asset yamls
+  into the new directory, generating the right kustomize / metadata yamls.
+
+* As we refactor and re-use the kustomize package used for loading user workloads,
+  add support for loading yamls from multiple directories, i.e.:
+  * `/usr/lib/microshift/manifests.d/00-user-app-1/`
+  * `/usr/lib/microshift/manifests.d/10-user-app-2/`
+  * `/etc/microshift/manifests.d/user-app-3/`
+  * `...`
+
+### Non-Goals
+
+* Making DNS component optional.
+
+* Making the certificate management component optional.
+
+* Allowing the replacement of network or storage components on runtime. Once
+  the system has booted to a storage or network component, the component cannot
+  be switched (but a full reset of the system `cleanup-all-microshift-data` and restart
+  should work for developers).
+
+## Proposal
+
+To allow modularity of the described core components the components are split across
+multiple microshift subpackages, supporting at least the following use cases:
+
+| Use case 	                                        | Packages installed in the RHEL4Edge image 	| Notes 	                            |
+|--------------------------------------------------	|-----------------------------------------  	|-------	                            |
+| The whole stack of components is needed         	| `microshift`                   	              | Nothing would change in this case, installing microshift, which becomes a virtual package, installs the necessary dependencies to help maintain backwards compatibility	|
+| Only advanced networking is needed                | `microshift-core`, `microshift-networking-ovnk`  |                                    	|
+| The bare minimum set of components is necessary   | `microshift-core`                             |                                   	|
+| Only persistent volumes are necessary             | `microshift-core`, `microshift-storage-topolvm` |                                     |
+
+
+The packages for networking and storage would be different in upstrean, containing
+references to unrestricted (OKD) container images, removing the need for a pull secret.
+
+The following changes would be introduced to allow the described type of modularity:
+
+Refactor and reuse the kustomize loader into a new module for core components which
+would load components in alphabetical order:
+
+`/usr/lib/microshift/components.d/005-networking-ovn/kustomize.yaml`
+`/usr/lib/microshift/components.d/010-storage-topolvm/kustomize.yaml`
```

Comment:
Would it make sense to add note that this alphabetical order also address the multus deployment issue (it needs to be after default cni plugin).

Reply:
Would multus go before or after ovn kubernetes? (to make an example)

Reply:
Done.

Reply:
Either way shall work. 
In openshift, multus is deployed before default CNI (e.g. ovnk), it contains a infinite loop in its entrypoint file to wait until the default CNI writes its CNI config file.
If multus is deployed after default CNI, it may not have to wait (if the default CNI pods can be guaranteed to become ready before multus).